curp/src/server/storage/wal/util.rs

```diff
@@ -0,0 +1,218 @@
+    }
+}
+
+impl Drop for LockedFile {
+    fn drop(&mut self) {
+        if self.file.is_some() && is_exist(self.path()) {
+            let _ignore = std::fs::remove_file(self.path());
+        }
+    }
+}
+
+/// Gets the all files with the extension under the given folder
+pub(super) fn get_file_paths_with_ext(
+    dir: impl AsRef<Path>,
+    ext: &str,
+) -> io::Result<Vec<PathBuf>> {
+    let mut files = vec![];
+    for result in std::fs::read_dir(dir)? {
+        let file = result?;
+        if let Some(filename) = file.file_name().to_str() {
+            if filename.ends_with(ext) {
+                files.push(file.path());
+            }
+        }
+    }
+    Ok(files)
+}
+
+/// Gets the parent dir
+pub(super) fn parent_dir(dir: impl AsRef<Path>) -> PathBuf {
+    let mut parent = PathBuf::from(dir.as_ref());
+    let _ignore = parent.pop();
+    parent
+}
+
+/// Fsyncs the parent directory
+pub(super) fn sync_parent_dir(dir: impl AsRef<Path>) -> io::Result<()> {
+    let parent_dir = parent_dir(&dir);
+    let parent = std::fs::File::open(parent_dir)?;
+    parent.sync_all()?;
+
+    Ok(())
+}
+
+/// Gets the checksum of the slice, we use Sha256 as the hash function
+pub(super) fn get_checksum(data: &[u8]) -> Vec<u8> {
+    let mut hasher = Sha256::new();
+    hasher.update(data);
+    hasher.finalize().into_iter().collect()
+}
+
+/// Validates the the data with the given checksum
+pub(super) fn validate_data(data: &[u8], checksum: &[u8]) -> bool {
+    get_checksum(data) == checksum
+}
+
+/// Checks whether the file exist
+pub(super) fn is_exist(path: impl AsRef<Path>) -> bool {
+    std::fs::metadata(path).is_ok()
+}
+
+/// Parses a u64 from u8 slice
+pub(super) fn parse_u64(bytes_le: &[u8]) -> u64 {
+    assert_eq!(bytes_le.len(), 8, "The slice passed should be 8 bytes long");
+    u64::from_le_bytes(
+        bytes_le
+            .try_into()
+            .unwrap_or_else(|_| unreachable!("This conversion should always exist")),
+    )
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{io::Read, process::Command};
+
+    use super::*;
+
+    #[test]
+    fn file_rename_is_ok() {
+        let mut tempdir = tempfile::tempdir().unwrap();
+        let mut path = PathBuf::from(tempdir.path());
+        path.push("file.test");
+        let lfile = LockedFile::open_rw(&path).unwrap();
+        let new_name = "new_name.test";
+        let mut new_path = parent_dir(&path);
+        new_path.push(new_name);
+        lfile.rename(new_name);
+        assert!(!is_exist(path));
+        assert!(is_exist(new_path));
+    }
+
+    #[test]
+    #[allow(clippy::verbose_file_reads)] // false positive
+    fn file_open_is_exclusive() {
+        let mut tempdir = tempfile::tempdir().unwrap();
+        let mut path = PathBuf::from(tempdir.path());
+        path.push("file.test");
+        let mut lfile = LockedFile::open_rw(&path).unwrap();
+        let path_str = path.to_str().unwrap();
+
+        let mut try_flock_output = Command::new("sh")
```

Comment:
Depend on the shell command may not be a good practice in test. Can we use `fcntl` to replace this `flock`?

Reply:
The shell command is removed.

Reply:
I initialy thought the flock is per-process lock, but it turns out it can be acquired independently in a single process.